src/main/java/jnr/unixsocket/UnixSocket.java

```diff
@@ -14,42 +17,144 @@
+
+class UnixSocket extends java.net.Socket {
+
+    private UnixSocketChannel chan;
+
+    private AtomicBoolean closed = new AtomicBoolean(false);
+    private AtomicBoolean indown = new AtomicBoolean(false);
+    private AtomicBoolean outdown = new AtomicBoolean(false);
+
+    private InputStream in;
+    private OutputStream out;
+
+    public UnixSocket(UnixSocketChannel chan) {
+        this.chan = chan;
+        in = Channels.newInputStream(chan);
+        out = Channels.newOutputStream(chan);
+    }
+
+    public void bind(SocketAddress addr) {
+        throw new UnsupportedOperationException("bind not supported");
+    }
+
+    @Override
+    public void close() throws IOException {
+        chan.close();
+        closed.compareAndSet(false, true);
+    }
+
+    @Override
+    public void connect(SocketAddress addr) throws IOException {
+        connect(addr, 0);
+    }
+
+    public void connect(SocketAddress addr, Integer timeout) throws IOException {
+        if (addr instanceof UnixSocketAddress) {
+            chan.connect((UnixSocketAddress) addr);
+        } else {
+            throw new IllegalArgumentException("address of type "
+                    + addr.getClass() + " are not supported. Use "
+                    + UnixSocketAddress.class + " instead");
+        }
+    }
+
+    @Override
+    public SocketChannel getChannel() {
+        return chan;
+    }
+
+    @Override
+    public InetAddress getInetAddress() {
+        return null;
+    }
 
-import java.nio.ByteBuffer;
-import java.nio.ByteOrder;
-import java.nio.channels.Channel;
+    public InputStream getInputStream() throws IOException {
+        if (chan.isConnected()) {
+            return in;
+        } else {
+            throw new IOException("not connected");
+        }
+    }
 
-public class UnixSocket {
-    private final NativeSocketChannel channel;
+    @Override
+    public SocketAddress getLocalSocketAddress() {
+        UnixSocketAddress address = chan.getLocalSocketAddress();
+        if (address != null) {
+            return address;
+        } else {
+            return null;
+        }
+    }
 
-    UnixSocket(NativeSocketChannel channel) {
-        this.channel = channel;
+    @Override
+    public OutputStream getOutputStream() throws IOException {
+        if (chan.isConnected()) {
+            return out;
+        } else {
+            throw new IOException("not connected");
+        }
     }
 
-    public final Channel getChannel() {
-        return channel;
+    @Override
+    public SocketAddress getRemoteSocketAddress() {
+        SocketAddress address = chan.getRemoteSocketAddress();
+
+        if (address != null) {
+            return address;
+        } else {
+            return null;
+        }
     }
 
-    public final void setKeepAlive(boolean on) {
-        Native.setsockopt(channel.getFD(), SocketLevel.SOL_SOCKET, SocketOption.SO_KEEPALIVE, on);
+    @Override
+    public boolean isBound() {
+        return false;
```

Comment:
Good catch! Yes of course. Previously, UnixSocketChannel did not support bind(), but now that we have bind(), this must be tracked with an AtomicBoolean in the channel and then be retrieved here. (Also: bind() fails to throw an AlreadyBoundException if the channel  is already bound)